### PR TITLE
Refactor classic menu conversion process

### PR DESCRIPTION
**Pull request by getdave**
## Description
Alternative approach to https://github.com/WordPress/gutenberg/pull/38824. Continues working towards addressing https://github.com/WordPress/gutenberg/issues/37190.

This PR seeks both to

- simplify the logic around converting classic menus to block based menus in the Nav block.
- avoid the impression of the UI "locking up" when the classic menu is selected and instead provide feedback to the user that an operation is taking place.

The current hooks are complex to reason about and don't expose enough information about the progress of the conversion to allow for the UI to respond (e.g. loading states...etc).

This PR builds on the approach outlined by @talldan in https://github.com/WordPress/gutenberg/pull/38824#pullrequestreview-883918504. Essentially we create a new hook which exposes:

1. A function which trigger the conversion process.
2. The state of the conversion process.

~🙏 **Please note**: This PR _does not_ attempt to solve the loading state indicator problems that are noticeable in this PR. This is being addressed in https://github.com/WordPress/gutenberg/pull/38907.~

## Outstanding problems to solve

- [x] Allow the dropdown to close immediately upon click to avoid the UI feeling like it's locking up.
- [x] ~Create loading UI states _in the block_ whilst the conversion process happens (good things we expose the state!).~ To be addressed following merge of https://github.com/WordPress/gutenberg/pull/38907.
- [x] Avoid component unmount errors when closing the dropdown whilst the async conversion process is still running.
- [x] Migrate towards usage of registry as suggest by @adamziel in https://github.com/WordPress/gutenberg/pull/38824#issuecomment-1041549542.
- [x] Handle errors from the async conversion process (`.catch()` .etc).
- [x] Provide accessible progress announcements of conversion status.

I will tackle those steps during the remainder of this week.

## Testing Instructions

1. Switch to a classic Theme and create a Menu using the `Menus` screen. Add lots of items just for fun.
2. Create another Menu with no items in it. Be sure to save.
3. Switch to block Theme.
4. New Post.
5. Insert a Navigation block.
6. Select the populated classic menu.
7. Immediately see a loading state appear and not sense that the UI is "locking up" (for an understand of what I mean check out the same proceedure on `trunk` or watch the screencaptures below).
8. See the menu converted successfully and inserted into the block.
9. Repeat the same process but this time using the unpopulated classic menu. You should still see the conversion happen but the block will end up being empty because the classic menu was empty. That's expected.
10. Try and test what happens if you go offline just before you click on the classic menu. The UI should handle things and you should see an error notice pop up.
11. Try using screen reader software to run the above steps. You should be able to perceive the progress of the conversion without having to look at the screen. It should be announced to you.

### Before

Notice how the dropdown doesn't close and  UI seems to lock up/freeze when I click the classic menu. 

https://user-images.githubusercontent.com/444434/154458859-495a22f5-4138-4dd9-87b1-55ca50706f2e.mp4

### After

Notice the UI lock up is gone. The dropdown disappears and we see the block loading state immediately which resolves as appropriate:

https://user-images.githubusercontent.com/444434/156602370-9478c9fe-c352-4bab-b137-affc5237ebad.mp4

#### Error handling

Notice that the UI communicates the error if the request fails:

https://user-images.githubusercontent.com/444434/156608077-385a689d-ebe2-43fc-a2f3-aa9ed90571d2.mov

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. 
- [ ] My code follows the accessibility standards. 
- [ ] I've tested my changes with keyboard and screen readers. 
- [ ] My code has proper inline documentation. 
- [ ] I've included developer documentation if appropriate. 
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal). 
- [ ] I've updated related schemas if appropriate.

### CODE REVIEW

File: packages/block-library/src/navigation/edit/use-convert-classic-menu-to-block-menu.js
Comment (adamziel) [inline packages/block-library/src/navigation/edit/use-convert-classic-menu-to-block-menu.js:28]: I'd pull it out of the hook, there's no need to construct a new function on each run and it would be more readable to me (more separation).

File: packages/block-library/src/navigation/edit/use-convert-classic-menu-to-block-menu.js
Comment (adamziel) [inline packages/block-library/src/navigation/edit/use-convert-classic-menu-to-block-menu.js]: Hey that's smart! I'd love to see a PR to `@wordpress/data` with this change.

Comment (adamziel) [reply packages/block-library/src/navigation/edit/use-convert-classic-menu-to-block-menu.js]: Just to make sure I understand what's going on – do you want to avoid errors when the action is dispatched by an unmounted component?

File: packages/block-library/src/navigation/edit/use-convert-classic-menu-to-block-menu.js
Comment (adamziel) [inline packages/block-library/src/navigation/edit/use-convert-classic-menu-to-block-menu.js:92]: You can have an async function inside useCallback, only you have to declare it separately, e.g.:

```js
const convert = useCallback(
	( menuId, menuName ) => {
		async function doTheWork() {};
		doTheWork();
	}
);
```

File: packages/block-library/src/navigation/edit/use-convert-classic-menu-to-block-menu.js
Comment (adamziel) [inline packages/block-library/src/navigation/edit/use-convert-classic-menu-to-block-menu.js]: That seems a bit overcomplicated to me for two state variables and three possible transitions, what do you think about the following?

```js
const [ navMenu, setNavMenu ] = useState( null );
const [ isFetching, setIsFetching ] = useState( false );
const setLoading = useCallback( () => {
	setIsFetching( true );
	setNavMenu( null );
} );
const setError = useCallback( () => {
	setIsFetching( false );
	setNavMenu( null );
} );
const setResolved = useCallback( ( menus ) => {
	setIsFetching( false );
	setNavMenu( menus );
} );
```

File: packages/block-library/src/navigation/edit/use-convert-classic-menu-to-block-menu.js
Comment (getdave) [reply packages/block-library/src/navigation/edit/use-convert-classic-menu-to-block-menu.js]: Yes. The "safe" aspect stops the dispatch if the component is unmounted. I will propose this in a core package in a separate PR/

**Comment by getdave** [comment]
I will resume work on this PR once https://github.com/WordPress/gutenberg/pull/38907 is merged.

File: packages/block-library/src/navigation/edit/use-convert-classic-menu-to-block-menu.js
Comment (getdave) [reply packages/block-library/src/navigation/edit/use-convert-classic-menu-to-block-menu.js:62]: Just noting I proposed this but it was rejected https://github.com/WordPress/gutenberg/pull/38956

File: packages/block-library/src/navigation/edit/use-convert-classic-menu-to-block-menu.js
Comment (getdave) [reply packages/block-library/src/navigation/edit/use-convert-classic-menu-to-block-menu.js]: Just noting I proposed this but it was rejected https://github.com/WordPress/gutenberg/pull/38956

File: packages/block-library/src/navigation/edit/index.js
Comment (getdave) [inline packages/block-library/src/navigation/edit/index.js]: This is confusing. Due to the `useState` var name containing the term `ref` It could appear we are setting a `React.ref` here which would not cause the component to re-render and thus the UI would not reflect the new menu ID.

In actual fact this is a `setState` call.

We should look to refactor this `ref` term away to something else.

File: packages/block-library/src/navigation/edit/use-convert-classic-menu-to-block-menu.js
Comment (getdave) [reply packages/block-library/src/navigation/edit/use-convert-classic-menu-to-block-menu.js:28]: I take your point but if you follow the flow you'll notice that this is required to be within a hook because it uses other hooks. In fact it's hooks within hooks within hooks.

I'm not in favour of this approach really but I don't think we should look to refactor in this PR because it will become overly complex to review.

How would you feel about pairing on a refactor of this sometime?

File: packages/block-library/src/navigation/edit/use-convert-classic-menu-to-block-menu.js
Comment (getdave) [reply packages/block-library/src/navigation/edit/use-convert-classic-menu-to-block-menu.js:92]: Yeh but in this case the aim isn't to return `Promise` from `convert()`. We want to run a routine and then have that routine update the `state` which we can then use outside of the hook to decide on various things such as loading states.

This relieves the hook consumer from having to reimplementing the `isFetching`, `isError`...etc inside their component. Rather all this is encapsulated and we simply expose a state which the consumer can use to decide on side effects within a `useEffect`.

File: packages/block-library/src/navigation/edit/use-convert-classic-menu-to-block-menu.js
Comment (getdave) [reply packages/block-library/src/navigation/edit/use-convert-classic-menu-to-block-menu.js]: I just felt the reducer was cleaner and easier to reason about. Moreover, if we want to avoid the set state on unmount error we're going to have to call the `useSafeDispatch` once for each `setState` in order to get the wrapped version of that handler.

File: packages/block-library/src/navigation/edit/index.js
Comment (getdave) [inline packages/block-library/src/navigation/edit/index.js:247]: Maybe the exposed state should be "converted" rather than "fetching". It would be more descriptive and we could do away with this var.

File: packages/block-library/src/navigation/edit/index.js
Comment (getdave) [inline packages/block-library/src/navigation/edit/index.js:260]: No showing a placeholder if we're running the conversion process.

File: packages/block-library/src/navigation/edit/index.js
Comment (getdave) [inline packages/block-library/src/navigation/edit/index.js:272]: Now `loading` UI state should show _immediately_ if we are running the conversion.

File: packages/block-library/src/navigation/edit/index.js
Comment (getdave) [inline packages/block-library/src/navigation/edit/index.js]: I don't love this. Open to ideas about how to improve.

File: packages/block-library/src/navigation/edit/index.js
Comment (getdave) [inline packages/block-library/src/navigation/edit/index.js]: Alternative to a separate `onSelectClassic` would be to detect whether it's a classic menu that's been selected by looking for a property of a classic menu that isn't present on a Navigation Menu (if there is one).

File: packages/block-library/src/navigation/edit/index.js
Comment (getdave) [inline packages/block-library/src/navigation/edit/index.js]: This kicks of the async process but is non blocking and allows the UI of the block to update independently based on the conversion status provided by the hook.

File: packages/block-library/src/navigation/edit/use-convert-classic-menu-to-block-menu.js
Comment (getdave) [inline packages/block-library/src/navigation/edit/use-convert-classic-menu-to-block-menu.js]: Centralised error handling here.

All errors are bubbled up to the UI via state changes on the reducer, which are then exposed via the hook's `state` property.

The consuming component can read `state.error` to get the error and `state.status` to determine if there was an error.

Note that `convertClassicMenuToBlockMenu` implements specialised error handling to ensure that error messages are descriptive of what has occurred (i.e. we don't rely on default messages coming from `core-data`...etc.

File: packages/block-library/src/navigation/edit/use-navigation-notice.js
Comment (getdave) [inline packages/block-library/src/navigation/edit/use-navigation-notice.js:13]: Allow for an optional custom message at call time

**Comment by getdave** [comment]
@jasmussen I wonder if you would take this for a spin on the design side?

**Comment by getdave** [comment]
Just spun this one out to capture the need to [avoid duplicate classic menu names during import](https://github.com/WordPress/gutenberg/issues/39200).

**Comment by getdave** [comment]
Accessibility review required just to validate the usage of `speak()` here is appropriate. I tested with voiceover and it seemed to provide a good level of feedback during the classic menu conversion process whereas before there was absolutely none.

File: packages/block-library/src/navigation/edit/use-convert-classic-menu-to-block-menu.js
Comment (adamziel) [reply packages/block-library/src/navigation/edit/use-convert-classic-menu-to-block-menu.js:28]: Sure, let's try to pair up tomorrow

File: packages/block-library/src/navigation/edit/use-convert-classic-menu-to-block-menu.js
Comment (adamziel) [reply packages/block-library/src/navigation/edit/use-convert-classic-menu-to-block-menu.js]: That error seems to be harmless and will disappear in the next React release though?

Comment (getdave) [reply packages/block-library/src/navigation/edit/use-convert-classic-menu-to-block-menu.js]: I've had a change of heart. Let's avoid adding complexity where it's not warranted. I've refactored to `useState`.

File: packages/block-library/src/navigation/edit/index.js
Comment (getdave) [reply packages/block-library/src/navigation/edit/index.js]: I've gone with the approach of detecting the classic menu by checking for the rather obscure `auto_add` prop.

File: packages/block-library/src/navigation/edit/use-convert-classic-menu-to-block-menu.js
Comment (adamziel) [reply packages/block-library/src/navigation/edit/use-convert-classic-menu-to-block-menu.js:92]: All I'm saying is there are two code styles available to choose from. Note that my snippet above doesn't return a promise either.

Comment (getdave) [reply packages/block-library/src/navigation/edit/use-convert-classic-menu-to-block-menu.js:92]: Oh I see. You're just saying we could move the async function outside of the `useCallback` and that would be ok.

Unless it's a blocker for you, I'd rather not do that now because I'm planning a follow up that might change things a little. I ping you when that's up.

File: packages/block-library/src/navigation/edit/index.js
Comment (scruffian) [inline packages/block-library/src/navigation/edit/index.js:350]: These messages are pretty helpful, should we also add them as snackbar notifications? (In a different PR)

**Review by scruffian — APPROVED** [review]
This works as described. The code looks good to me, but I'm not familiar with it.

File: packages/block-library/src/navigation/edit/index.js
Comment (getdave) [reply packages/block-library/src/navigation/edit/index.js:350]: We could yes if that's not too "noisy"? I wonder if notification popups have `speak()` built in?

Comment (getdave) [reply packages/block-library/src/navigation/edit/index.js:350]: Actually I've found a great case where that would be very useful. When creating an empty menu it's not clear that it's been "successful". An explicit confirmation would be really helpful.

I'm working on a sister PR now which I will push up which refactors the create empty process.

File: packages/block-library/src/navigation/edit/use-convert-classic-menu-to-block-menu.js
Comment (adamziel) [inline packages/block-library/src/navigation/edit/use-convert-classic-menu-to-block-menu.js]: I think this information is lost. [The second argument to Error is `options`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Error/Error#rethrowing_an_error_with_a_cause) and the only acceptable option is `cause`. You could wrap that in `cause` as follows:

```suggestion
			{
				cause: {
					menuId,
					menuName,
				}
			}
```

But it doesn't seem like we're actually doing anything with this information so we could just remove it entirely:

```suggestion
```

Also, I believe the stack trace points to the line where the error is constructed, not thrown. How about reusing a variable called `errorMessage = sprintf( // ...` instead of `fetchError`?

File: packages/block-library/src/navigation/edit/use-convert-classic-menu-to-block-menu.js
Comment (adamziel) [inline packages/block-library/src/navigation/edit/use-convert-classic-menu-to-block-menu.js]: This could be deleted, too

File: packages/block-library/src/navigation/edit/use-convert-classic-menu-to-block-menu.js
Comment (adamziel) [inline packages/block-library/src/navigation/edit/use-convert-classic-menu-to-block-menu.js]: This is only used once – let's inline it.

File: packages/block-library/src/navigation/edit/use-convert-classic-menu-to-block-menu.js
Comment (adamziel) [inline packages/block-library/src/navigation/edit/use-convert-classic-menu-to-block-menu.js]: We lose the error information here. How about `throw new Error( fetchMessage, { cause: e } );`?

File: packages/block-library/src/navigation/edit/use-convert-classic-menu-to-block-menu.js
Comment (adamziel) [reply packages/block-library/src/navigation/edit/use-convert-classic-menu-to-block-menu.js]: Or better, replaced with { cause: e }

File: packages/block-library/src/navigation/edit/use-convert-classic-menu-to-block-menu.js
Comment (adamziel) [inline packages/block-library/src/navigation/edit/use-convert-classic-menu-to-block-menu.js]: Let's re-throw this exception so that it can be debugged in devtools.

File: packages/block-library/src/navigation/edit/use-convert-classic-menu-to-block-menu.js
Comment (adamziel) [reply packages/block-library/src/navigation/edit/use-convert-classic-menu-to-block-menu.js:28]: With the recent updates in place I feel pretty good about having this here, let's just wrap it in `useCallback`.

File: packages/block-library/src/navigation/edit/index.js
Comment (adamziel) [inline packages/block-library/src/navigation/edit/index.js]: We use `"pending"` here and `CLASSIC_MENU_CONVERSION_PENDING` in the other file – how about using these constants here as well? I don't like that long name, though. Ideally we could rely on TS to verify the right-hand side here.

**Comment by getdave** [comment]
Thanks for catching (excuse the pun) those lost errors @adamziel. Let me rework those.

File: packages/block-library/src/navigation/edit/index.js
Comment (getdave) [reply packages/block-library/src/navigation/edit/index.js]: Yeh I was thinking that as well 😄 

AFAIK we aren't supposed to use TS at this level (Gutenberg informal agreement) but yes it would be ideal.

File: packages/block-library/src/navigation/edit/use-convert-classic-menu-to-block-menu.js
Comment (getdave) [reply packages/block-library/src/navigation/edit/use-convert-classic-menu-to-block-menu.js]: Just checking...if we rethrow the error and it isn't handled then won't the block crash? I was trying to avoid that.

File: packages/block-library/src/navigation/edit/use-convert-classic-menu-to-block-menu.js
Comment (getdave) [reply packages/block-library/src/navigation/edit/use-convert-classic-menu-to-block-menu.js]: `menuItemsParameters` is now inlined.

File: packages/block-library/src/navigation/edit/use-convert-classic-menu-to-block-menu.js
Comment (getdave) [reply packages/block-library/src/navigation/edit/use-convert-classic-menu-to-block-menu.js]: ✅ I've rethrown the error and avoided overthinking error abstractions.

File: packages/block-library/src/navigation/edit/use-convert-classic-menu-to-block-menu.js
Comment (getdave) [reply packages/block-library/src/navigation/edit/use-convert-classic-menu-to-block-menu.js]: I did see this `cause` and was thinking "I'll come back to that". Guess what...I forgot! Lesson - don't leave issues/problems undocumented as you come across them - you will forget 😄 

Thanks for catching the silly usage of `options`. I've now updated to pass the original error as `cause`.

File: packages/block-library/src/navigation/edit/use-convert-classic-menu-to-block-menu.js
Comment (getdave) [reply packages/block-library/src/navigation/edit/use-convert-classic-menu-to-block-menu.js]: I've ended up inlining everything and passing the original error as `cause`.

File: packages/block-library/src/navigation/edit/use-convert-classic-menu-to-block-menu.js
Comment (getdave) [reply packages/block-library/src/navigation/edit/use-convert-classic-menu-to-block-menu.js]: Ok I tested and it's all fine. The block handles the error but the message is still displayed in dev tools. Win 👍

File: packages/block-library/src/navigation/edit/index.js
Comment (getdave) [reply packages/block-library/src/navigation/edit/index.js]: ✅

File: packages/block-library/src/navigation/edit/use-convert-classic-menu-to-block-menu.js
Comment (adamziel) [inline packages/block-library/src/navigation/edit/use-convert-classic-menu-to-block-menu.js:111]: nit, this could say just `throw err` since we're not going to tell the user what the error message says.

**Comment by adamziel** [comment]
The only remaining thing I don't like is how many times it says _classic menu conversion_. I don't want to block on that, though, especially since I don't have any shorter naming scheme to propose.


**Edit** actually, I left a few more notes

File: packages/block-library/src/navigation/edit/index.js
Comment (adamziel) [reply packages/block-library/src/navigation/edit/index.js]: If it made sense to say that in the PR, it makes sense to say it in a comment in that file

File: packages/block-library/src/navigation/edit/index.js
Comment (adamziel) [inline packages/block-library/src/navigation/edit/index.js]: Could that be a separate function? My reasoning:

* this logic is repeated in two places
* I'd rather have the future us find that function when we need it, than miss this very specific part of edit/index.js and invent another way of checking whether something is a classic menu

File: packages/block-library/src/navigation/edit/index.js
Comment (adamziel) [inline packages/block-library/src/navigation/edit/index.js]: Can `navPostOrClassicMenu` be null? If yes, then the hasOwnProperty call above should take that into account.

File: packages/block-library/src/navigation/edit/index.js
Comment (adamziel) [inline packages/block-library/src/navigation/edit/index.js]: Ditto the other two comments

File: packages/block-library/src/navigation/edit/index.js
Comment (getdave) [reply packages/block-library/src/navigation/edit/index.js]: For now I've exited early. Perhaps we ought to consider showing a error, but I can't yet see a way where this would be `null` although it might happen.

File: packages/block-library/src/navigation/edit/index.js
Comment (getdave) [reply packages/block-library/src/navigation/edit/index.js]: Yes I agree. Looks like overtime the signatures of this handler become identical. That's good news as the incremental refactoring is obviously making the codebase easier to reason about.

I've extracted a common utility.

**Comment by getdave** [comment]
> The only remaining thing I don't like is how many times it says _classic menu conversion_

I agree. I think perhaps a var naming follow up is in order. Especially when you see which way https://github.com/WordPress/gutenberg/pull/39219 is heading.

File: packages/block-library/src/navigation/edit/index.js
Comment (adamziel) [inline packages/block-library/src/navigation/edit/index.js]: nit: `?` not needed since we return early

File: packages/block-library/src/navigation/edit/index.js
Comment (adamziel) [inline packages/block-library/src/navigation/edit/index.js]: ditto

File: packages/block-library/src/navigation/edit/index.js
Comment (getdave) [reply packages/block-library/src/navigation/edit/index.js]: ✅

File: packages/block-library/src/navigation/edit/index.js
Comment (getdave) [reply packages/block-library/src/navigation/edit/index.js]: ✅

**Comment by getdave** [comment]
I'm going to merge this once tests pass unless I get any more ❌ feedback.

**Comment by jasmussen** [comment]
I saw the ping last week and did not have time to look at it. From a quick glance it looks good:

nav

The following issue is present in trunk, and subject to separate conversations, so not something that should block this PR. But showing the old menu even after it's being converted, as well as tools to manage them, it adds a fair bit of confusion: when do those tools appear? Why can I convert my menu again, wasn't it converted already? Worth us all looking at separately, but not precluded by these enhancements. Nice!

**Comment by getdave** [comment]
> I saw the ping last week and did not have time to look at it. 

No worries.

> From a quick glance it looks good

👍 


> The following issue is present in trunk, and subject to separate conversations, so not something that should block this PR. But showing the old menu even after it's being converted, as well as tools to manage them, it adds a fair bit of confusion: when do those tools appear? Why can I convert my menu again, wasn't it converted already? Worth us all looking at separately, but not precluded by these enhancements. Nice!

Agreed. I feel like there's an Issue tracking this but I can't find it right now.

**Pull request merged**
